### PR TITLE
feat(ios): partial conversion of internal logs to use of Sentry

### LINKED ISSUE #3474: chore(ios): Error reporting/logging rework
As noted on #3470, which saw the introduction of more explicit error definitions:

> All Keyman-generated errors now have defined types and/or enums devoted to their maintenance, along with localized messages. Some errors also accept parameters, though these aren't passed along to the messages. The parameters would be far more useful to programmers (who can condition on the enums and extract said parameters) than end-users, who want a simple, minimally-confusing localized message instead.

There are a number of non-fatal errors currently handled within `catch` blocks that may provide us with useful information for iOS maintenance.  We should log these with Sentry, rather than constantly relying on a lot of `log.info` / `log.error` / etc.

**Important note**:  Updating our Sentry dependency from 4.4.x to 5.2.0 at the same time will greatly facilitate this, as it adds methods that can receive error objects... unlike 4.4.x.  It also brings the Sentry API up to parity with the rest of their "unified API" clients, which should make cross-platform maintenance/reviews of Sentry-related code easier for the team.

### PR DESCRIPTION
**Pull request by jahorton**
In service of #3474.

Dependent upon #5120.

While not an exhaustive conversion of all our internal logging messages, this converts many of our iOS platform's relatively-simple and important log messages into Sentry events.  I inspected each as I passed through them; some events receive mildly sanitized data as a result.  `captureAndLog` statements in here should _not_ occur during standard use of the app and are things we'd want to know about should they occur.

Many of the `log.error` statements left unconverted in the codebase regard trickier scenarios.  Some occur in relatively common cases - such as the app's first load, before certain properties may initialize.  Other have interactions with internet connectivity - errors may result in the keyboard search should the connection drop, and that's a scenario that should be fully handled within the app and not reported to Sentry.  Finally, some deal with parts I'm less up-to-speed on, such as font management.

--------

Noting some of the new breadcrumbs this provides, it may be worth our while to add a new settings option to generate a manual "error report".  That could trigger a custom Sentry event with user feedback that would also send along all the breadcrumbs, which could be used to provide useful information and serve as a more easily-accessible "log".  Though... it would probably only help in-app, so... not sure how useful that'd be.

Thinking about it... this is probably enough to "fix #3474", though we'll want to create a secondary issue to address the harder-to-convert `log.info` and `log.error` entries at some point in the future.  They'll require considerably more time investment to do well, so those ought be triaged as "future" work.

### CODE REVIEW

File: ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift [hunk 1]
```diff
@@ -50,6 +51,7 @@ public class SentryManager {
         return nil
       }
     }
+    //
```
Comment (darcywong00) [inline ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift:54]: Is this a TODO for something?

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift [hunk 2]
```diff
@@ -283,15 +285,28 @@ extension KeymanWebViewController {
     do {
       data = try JSONSerialization.data(withJSONObject: stub, options: [])
     } catch {
-      log.error("Failed to serialize keyboard stub: \(error)")
+      let event = Sentry.Event(error: error)
+      event.message = SentryMessage(formatted: "Failed to serialize keyboard stub: \(error)")
+      event.extra = [:]
+      event.extra!["id"] = stub["KI"]
+      event.extra!["package"] = stub["KP"]
```
Comment (darcywong00) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift:292]: Can stub["KP"] be undefined?

Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift:292]: `undefined` does not exist in Swift.  The worst case is `nil`, and that wouldn't break this.

File: ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift [hunk 3]
```diff
@@ -89,6 +91,107 @@ public class SentryManager {
...
+    crumb.level = sentryLevel
+    if let category = category {
+      crumb.category = category
+    }
+
+    self.breadcrumbAndLog(crumb: crumb, logLevel: logLevel)
+  }
+
   public static func forceError() {
     SentrySDK.addBreadcrumb(crumb: Sentry.Breadcrumb(level: .info, category: "Deliberate testing error"))
```
Comment (darcywong00) [inline ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift:196]: Should these also be in
```
if _started {
}
```

Granted, this is outside the scope of the PR

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift [hunk 4]
```diff
@@ -283,15 +285,28 @@ extension KeymanWebViewController {
...
+
+      SentryManager.captureAndLog(event)
       return
     }
 
-    log.debug("Keyboard stub: \(stubString)")
+    SentryManager.breadcrumbAndLog("Keyboard stub built for \(keyboard.id)", logLevel: XCGLogger.Level.none)
+    log.info("Keyboard stub: \(stubString)")
```
Comment (darcywong00) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift:309]: Why not concatenate with breadcrumbAndLog above? (same for l. 347)

File: ios/engine/KMEI/KeymanEngine/Classes/Settings/InstalledLanguagesViewController.swift [hunk 5]
```diff
@@ -173,7 +173,9 @@ public class InstalledLanguagesViewController: UITableViewController, UIAlertVie
       if let langName = keyboardLanguages[l]?.name {
         log.info("keyboard language \(l) \(langName) has lexical model")
```
Comment (darcywong00) [inline ios/engine/KMEI/KeymanEngine/Classes/Settings/InstalledLanguagesViewController.swift:174]: Do you want to breadcrumbAndLog here?

File: ios/engine/KMEI/KeymanEngine/Classes/Settings/SettingsViewController.swift [hunk 6]
```diff
@@ -399,7 +400,9 @@ open class SettingsViewController: UITableViewController {
       if let langName = keyboardLanguages[l]?.name {
         log.info("keyboard language \(l) \(langName) has lexical model")
```
Comment (darcywong00) [inline ios/engine/KMEI/KeymanEngine/Classes/Settings/SettingsViewController.swift:401]: breadcrumb?

File: ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift [hunk 7]
```diff
@@ -89,6 +91,107 @@ public class SentryManager {
...
+   * Will safely bypass the Sentry component if not activated by the app, only logging the
+   * message in such scenarios.
+   */
+  public static func captureAndLog(_ event: Sentry.Event, logLevel: XCGLogger.Level? = nil) {
+    // Guarded in case a library consumer decides against initializing Sentry.
+    if _started {
+      SentrySDK.capture(event: event)
+    }
+
+    let level = logLevel ?? mapLoggingLevel(event.level)
```
Comment (darcywong00) [inline ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift:126]: You've can handle event.info, but I see you've replaced all the 
`log.info` with `breadcrumbAndLog`.

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift (hunk 4 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift:309]: Here, `log.info` is printing the _whole stub_, not just the keyboard's ID.  I'm not sure that we should Sentry-log the entire stub.

File: ios/engine/KMEI/KeymanEngine/Classes/Settings/InstalledLanguagesViewController.swift (hunk 5 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Settings/InstalledLanguagesViewController.swift:174]: I felt that the two spots with this particular check were a bit 'noisy' and not particularly informative, so I skipped 'em.

**Review by mcdurdin — APPROVED** [review]

File: ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift (hunk 3 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift:197]: The functions they call already perform that filtering.  Doing so here would actually prevent the local 'log' statements from executing, and we don't want those blocked.

It's possible a library user may not wish to utilize Sentry.  In such cases, we should still make sure the on-device logs receive any error message sent through this helper.

File: ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift (hunk 7 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Errors/SentryManager.swift:126]: There are a few spots with `log.debug` for breadcrumbs, and there's at least one case coming with a `.fatal` sentry-level error instead of plain `.error`.

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.58-alpha](https://keyman.com/downloads/releases/alpha/15.0.58)
